Add `read_only` field setting

joshua-sharp opened the pull request:
If present and set to true, the field will be rendered on forms but
with a 'readonly' attribute set. It will also be greyed out. This is
intended for use with fields that should be visible to users, but not
updated. An example would be fields storing IDs that link entities to
external systems. It would be useful for users to be able to see these,
but not update them as this could cause inconsistencies.

Once field-level permissions have been added, this feature may become
redundant. Fields that the user has permission to view but not edit
could automatically be disabled and greyed out.

An example field definition:
```
{
  "title": "External ID",
  "identifier": "external_id",
  "type": "text",
  "settings": {
    "read_only": true
  }
}
```

joshua-sharp commented:
A notable change since the last time this feature was PRed is that read-only fields use the `readonly` HTML attribute, instead of `disabled`. This is to ensure the field data is submitted, which doesn't happen when using `disabled`.

franzwilding commented:
Thank you for the PR, nice work. However, as far as I see it, you only add an HTML attribute and do not check on the server side if the value was modified. 

Also, in many cases, the value of read_only depends on the current user or other context variables.
As I will implement a field permissions #417 feature any time soon, I would like to merge part of your PR into this feature. It could work like this: 

- When a user is allowed to view a field but not to update it, we set the disabled option instead of the  read_only option to the form element (and pass it to the Vue elements, using your code). 
- In any case, if the user is not allowed to update the field, we skip this field when updating the content. Because we do this, it is no problem at all, that a disabled Symfony form element will not get passed to the server.

What do you say?

joshua-sharp commented:
Thanks, that sounds like a good idea. Having server-side validation of read-only fields would definitely be an improvement, and it ties into the field-level permission system you mentioned.

One point in favour of `readonly` is that it allows users to select and copy the contents of the field. `disabled` fields don't allow their content to be selected, and we found it helpful for users to be able to easily copy values from read_only fields.

I'd suggest rendering read_only fields in a way that allows selection their contents. The field contents could be shown as a non-input text element, or a dummy `readonly` field with no `name` could be rendered.

franzwilding commented:
> I'd suggest rendering read_only fields in a way that allows selection their contents. The field contents could be shown as a non-input text element, or a dummy `readonly` field with no `name` could be rendered.

You are right! I will come back to you when I start working on the field permissions

franzwilding commented:
Hey @joshua-sharp ! Field permissions are now implemented (LIST, VIEW, UPDATE). 
At the moment. when the user is not allowed to UPDATE the field, it will be removed before the form is built. 

With your change, we could instead render the field but as read_only, allowing to show / select the value but not to change it.

joshua-sharp commented:
Hi @franzwilding 

I've implemented the changes to make fields read-only if the user has the VIEW field permission but not the LIST and UPDATE permissions. The read_only flag is now only used internally while the form is being built, and can't be set from the domain JSON.

As for creating tests for this feature, I've removed the read_only setting assertions in each field type's tests as they are no longer relevant. I'm unsure how to test whether the read_only setting is being updated correctly after building the form, as the authorizationChecker is always granting access to VIEW, UPDATE and LIST fields from within a test function.

Could you please advise if any changes should be made to the implementation, and what direction should be taken for testing this feature?
